### PR TITLE
Add: sqlglot to formalize sql clause

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

- Add sqlglot transpiler to parse the sql command before any SQL execution
- Fix the `--with` parameter with double syntax issue on mysql (#145)
- Fix the `sqlplot histogram` [with double quote syntax issue](https://github.com/ploomber/jupysql/pull/152#issuecomment-1442805171) on mysql 

## Issue ticket number and link

Close https://github.com/ploomber/jupysql/issues/158 - Integrate with sqlglot
Close https://github.com/ploomber/jupysql/issues/145 - `--with` parameter with double syntax issue on mysql

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [x] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--164.org.readthedocs.build/en/164/

### CODE REVIEW

File: src/sql/connection.py [hunk 1]
```diff
@@ -348,3 +353,17 @@ def _get_curr_connection_info(self):
             "driver": getattr(engine.dialect, "driver", None),
             "server_version_info": getattr(engine.dialect, "server_version_info", None),
         }
+
+    def _transiple_query(self, query):
```
Comment (Mause) [inline src/sql/connection.py:357]:
```suggestion
    def _transpile_query(self, query):
```

File: src/sql/connection.py [hunk 2]
```diff
@@ -348,3 +353,17 @@ def _get_curr_connection_info(self):
...
+        connection_info = self._get_curr_connection_info()
+        try:
+            write_dialect = connection_info["dialect"]
+            if write_dialect in DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING:
+                write_dialect = DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING[
+                    write_dialect
+                ]
```
Comment (Mause) [inline src/sql/connection.py:364]:
```suggestion
            write_dialect = DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING.get(write_dialect, write_dialect)
```

File: src/sql/connection.py [hunk 3]
```diff
@@ -348,3 +353,17 @@ def _get_curr_connection_info(self):
...
+            query = sqlglot.transpile(query, read="duckdb", write=write_dialect)[0]
+        except Exception as e:
+            raise e
```
Comment (Mause) [inline src/sql/connection.py:367]:
```suggestion
```

**Comment by idomic** [comment]
@edublancas I think it somehow closed the PR?

File: src/sql/connection.py (hunk 3 above)
Comment (tonykploomber) [reply src/sql/connection.py]: thx, now it's fix, it's cleaner, good catch!

File: src/sql/run.py [hunk 4]
```diff
@@ -400,6 +401,8 @@ def run(conn, sql, config, user_namespace):
                 )[0]
                 result = FakeResultProxy(cur, headers)
             else:
+                if Connection.current:
+                    statement = str(Connection.current._transpile_query((statement)))
```
Comment (tonykploomber) [inline src/sql/run.py:405]: @edublancas 
Let me know if you think here is the right place to parse the original SQL clause with sqlglot

File: src/sql/connection.py [hunk 5]
```diff
@@ -42,6 +43,10 @@
     "pymssql": "pymssql",
 }
 
+DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING = {
```
Comment (tonykploomber) [inline src/sql/connection.py:46]: @edublancas 
The reason to have this mapping is the dialect name isn't fully consistent between sqlglot & sqlaclchemy, need this map to formalize the dialect name

File: src/sql/connection.py [hunk 6]
```diff
@@ -348,3 +353,13 @@ def _get_curr_connection_info(self):
...
+
+    def _transpile_query(self, query):
+        connection_info = self._get_curr_connection_info()
+        try:
+            write_dialect = DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING.get(
+                connection_info["dialect"], connection_info["dialect"]
+            )
+            query = sqlglot.transpile(query, read="duckdb", write=write_dialect)[0]
+        finally:
+            return query
```
Comment (edublancas) [inline src/sql/connection.py:365]: why do we need this try ... finally?

File: src/sql/plot.py [hunk 7]
```diff
@@ -283,7 +279,8 @@ def _min_max(con, table, column, with_=None):
 
     if with_:
         query = str(store.render(query, with_=with_))
-
+    if sql.connection.Connection.current:
```
Comment (edublancas) [inline src/sql/plot.py:282]: what happens if we remove this if statement? is there any case where there isn't an active connection and the user is able to execute this?

File: src/sql/run.py (hunk 4 above)
Comment (edublancas) [reply src/sql/run.py:405]: this would transplite the user's query right? e.g.

```
%%sql
select * from table
```

why do we transpile user's queries?

I'd assume we only need to transpile queries that we run on user's behalf that they don't write explicitly (e.g., when plotting), but maybe I'm missing something

File: src/tests/integration/conftest.py [hunk 8]
```diff
@@ -103,11 +103,18 @@ def load_taxi_data(engine):
     df.to_sql(name=table_name, con=engine, chunksize=100_000, if_exists="replace")
 
 
+def load_plot_data(engine):
+    table_name = "plot_something"
+    df = pd.DataFrame({"x": range(0, 5), "y": range(5, 10)})
+    df.to_sql(name=table_name, con=engine, chunksize=100_000, if_exists="replace")
```
Comment (edublancas) [inline src/tests/integration/conftest.py:109]: let's make this dataset smaller, say 1000 observations.  there's no need to make it this big, it'll only slow things down during testing

File: src/tests/integration/test_generic_db_opeations.py [hunk 9]
```diff
@@ -140,3 +152,70 @@ def test_telemetry_execute_command_has_connection_info(
...
+
+@pytest.mark.parametrize(
+    "cell",
+    [
+        ("%sqlplot histogram --table plot_something --column x"),
+        ("%sqlplot hist --table plot_something --column x"),
+        ("%sqlplot histogram --table plot_something --column x --bins 10"),
+        pytest.param(
+            "%sqlplot boxplot --table plot_something --column x",
+            marks=pytest.mark.xfail(reason=BOX_PLOT_FAIL_REASON),
```
Comment (edublancas) [inline src/tests/integration/test_generic_db_opeations.py:170]: this is xfailing the tests with all databases right? can just xfail with MySQL? (I think all the others have percentile_disc)

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/connection.py (hunk 6 above)
Comment (tonykploomber) [reply src/sql/connection.py:365]: Since `sqlglot` might throw the Exception when the write_dialect name isn't supported, e.g. `sqlglot.transpile(query, read="duckdb", write="postgresql")[0]`, I am thinking when the write_dialect isn't support by the `sqlglot`, we can still safely return the original query

File: src/sql/run.py (hunk 4 above)
Comment (edublancas) [reply src/sql/run.py]: I think let's move the transpile step [here](https://github.com/ploomber/jupysql/blob/7b9f0f17c41342c31872708dd431633c7576b09b/src/sql/store.py#L86); I think the problem is that the expected query in the test case is hardcoded; hence it might not match.

there are some changes we need to do here since now SQLStore must be aware of the current dialect (so it can pass it to SQLQuery and then, pass it to the transpile call).

File: src/sql/connection.py (hunk 6 above)
Comment (edublancas) [reply src/sql/connection.py:365]: ah. makes sense. let's add some telemetry when transpile fails

File: src/sql/plot.py (hunk 7 above)
Comment (tonykploomber) [reply src/sql/plot.py:282]: Handled in this PR
https://github.com/ploomber/jupysql/pull/171

Let's merge that then we can remove here

File: src/tests/integration/test_generic_db_opeations.py (hunk 9 above)
Comment (tonykploomber) [reply src/tests/integration/test_generic_db_opeations.py:170]: Fix by separating `boxplot` and `histogram` test case

**Comment by edublancas** [comment]
looks like the docs are failing, my guess is that this is due to the new jupyter book version: https://github.com/executablebooks/jupyter-book/releases

let's pin to `<0.14.0` and see if that fixes it

**Comment by tonykploomber** [comment]
Build is good now!
@edublancas

File: src/sql/plot.py (hunk 7 above)
Comment (edublancas) [reply src/sql/plot.py:282]: cool, I reviewed #171, let's merge that one first then

File: src/tests/integration/test_generic_db_opeations.py (hunk 9 above)
Comment (edublancas) [reply src/tests/integration/test_generic_db_opeations.py:170]: perhaps I'm missing something but seems like we're still marking xfail for all databases? we should check that we can create the boxplot with every database that has the percentile disc, AFAIK the only one that doesn't implement it is MySQL

File: src/sql/store.py [hunk 10]
```diff
@@ -83,9 +83,12 @@ def __init__(self, store: SQLStore, query: str, with_: Iterable = None):
 
     def __str__(self) -> str:
         with_all = _get_dependencies(self._store, self._with_)
-        return _template.render(
+        rendered_template = _template.render(
             query=self._query, saved=self._store._data, with_=with_all
         )
+        if sql.connection.Connection.current:
+            return sql.connection.Connection.current._transpile_query(rendered_template)
```
Comment (edublancas) [inline src/sql/store.py:90]: we have an edge case where. because the SQLStore is combining SQL code from two sources. Given a CTE:

```sql
with something as (
    select * from ...
), another as (
    select * from ...
)
select * from ...
```

we're writing the CTEs (`with something as (...)`) on behalf of the user, but the user is supplying the inner `select * from ...` queries. So I think we need a bit to tweak the logic a bit here. we need to transpile the [template](https://github.com/ploomber/jupysql/blob/ecd91c1a598a2c9f1beeecf5c17474795b64a574/src/sql/store.py#LL66) by using duckdb as the source dialect and the user's dialect as target, and then, render the template with jinja by adding the `select * from` statements unchanged since they're supplied by the user and we should not transpile user-submitted code - let's add a test case on this

Comment (tonykploomber) [reply src/sql/store.py:90]: I have some feedbacks.

### Here is the plan:
1. Use sqlglot to transpile the template, which contains some unparsed raw sql clauses supplied by user, something like `WITH "with_query" AS ({{save_query_with_query}}) {{user_query}}`
2. After it's transpiled, use jinja to populate the raw sql clauses supplied by user

However, I was getting some error on step one, when I executed:
```
query = """
WITH with_query AS ({{save_query_with_query}}) {{user_query}}
"""
print(sqlglot.transpile(query, read="duckdb", write="mysql")[0])
```
Shows error, seems we can't transpile the query clause with {{variable}} 
```
ParseError: struct does not support CTE.
  
WITH with_query AS ({{save_query_with_query}}) {{user_query}}
```

Example of query clause without {{variable}}, this executes fine
```
query = """
WITH with_query AS (SELECT * FROM test) SELECT * from with_query
"""
print(sqlglot.transpile(query, read="duckdb", write="mysql")[0])
```


## Another possible workaround solution
https://sqlglot.com/sqlglot.html#build-and-modify-sql

We might write the sql by using sqlglot expression, then parse to the target dialect, which might be a more generic way to define our internal sql clause (even possible to solve the problem with [PercentileDisc](https://sqlglot.com/sqlglot/expressions.html#PercentileDisc))

I am still doing some research on this

Comment (tonykploomber) [reply src/sql/store.py]: ### Update - workaround solution
The [latest solution is here](https://github.com/ploomber/jupysql/pull/164/files#diff-a3d1291b54f0d59f191101a81a116bad5859a2bd81603b9b7c1cbea77d18cc10R74-R81), I abandoned the jinja template method since we can natively construct the query with saved `with clause`

So we want to construct the final query with some predefined `with clause` like this:
```
WITH sub_query_a AS (SELECT * FROM table_a) 
WITH sub_query_b AS (SELECT * FROM table_b) 
SELECT * FROM sub_query_a 
```

The constraint is we want to "keep" all user-supplied queries untouched, so translate to the sqlglot fashion will be like this
`parse_one("SELECT * FROM sub_query_a").with_("sub_query_a", "SELECT * FROM table_a").with_("sub_query_b", "SELECT * FROM table_b").sql()`


I think this way is the best way to keep all user-supplied statements, since we already use the native way to construct the user query with `with clause`. If we provide the dialect attribute in `parse_one` or `with_`, it might transpile the user statement though

Passed the integration testing among all databases, but failed on the doc build now

Ref: I asked [the question](https://tobiko-data.slack.com/archives/C0448SFS3PF/p1677792457879019) in sqlglot slack channel

Comment (edublancas) [reply src/sql/store.py]: nvm. we still need jinja for the new variable substitution format

**Comment by tonykploomber** [comment]
@edublancas 

TL;DR;
If we gonna use sqlglot to implement this, we can't allow the --with name contains - (dash symbol)

### Issue

Something I noticed that we are using [with_](https://sqlglot.com/sqlglot/expressions.html#Subqueryable.with_) to construct the query with subqueries, however, the API itself doesn't allow `alias` contains `-` (dash symbol)

E.G. This will be fail
```
from sqlglot import parse_one
parse_one("SELECT * FROM sub_query_a").with_("with-q", "SELECT * FROM test").sql()
```

This issue is from the build doc CI/CD flow, which we have magic-plot.md that defines
```
%sqlplot histogram --table no-nulls --column body_mass_g --with no-nulls
```

The `no-nulls` contains `-` (dash symbol), causing the doc build fail.

### Tried to workaround

I tried to use [to_identifier](https://sqlglot.com/sqlglot/expressions.html#to_identifier) to quote the alias, but it will fail on mysql and mariadb, since those database doesn't support the quote syntax

Ref: [Discussion in sqlglot thread](https://tobiko-data.slack.com/archives/C0448SFS3PF/p1677866766156519)

### What  can do if we going to use sqlglot

When the user uses `--with` with name contains `-`, we just disallow it, since the sqlglot can't handle this, probably throw some error message to suggest user not to use `-` name on save query

**Comment by tobymao** [comment]
```
Tried to workaround

I tried to use [to_identifier](https://sqlglot.com/sqlglot/expressions.html#to_identifier) to quote the alias, but it will fail on mysql and mariadb, since those database doesn't support the double quote query like 'WITH "with-q" AS (SELECT * FROM test) SELECT * FROM sub_query_a'
```

you're probably not doing something right. if you specify mysql / mariadb, then it will translate the double quote to a back tick.

**Comment by tonykploomber** [comment]
> > If we gonna use sqlglot to implement this, we can't allow the --with name that contains - (dash symbol)
> 
> Can we normalize the query?
> 
> Eg: the user passes --with no-nulls. We normalize it as no_nulls (so the CTE doesn't break). But the user can still refer to it with no-nulls.
> 
> I think we can emit a warning to tell the user what we'll normalize it.


Normalize is possible, but it's kinda tricky when it comes to practice

Since not only the table name probably needs normalize, the user query might need to normalize as well

e.g.

```
%sql --save taxi-subset --no-execute SELECT * FROM taxi LIMIT 3
%sql --with taxi-subset SELECT * FROM taxi-subset
```

The `SELECT * FROM taxi-subset` will need to normalize to `SELECT * FROM taxi_subset` before using `sqlglot.parse_one`, we might able to use something like `string.replace('-', '_')` to normalize, but it's also possible to break the query statement that we don't want normalize it, e.g. `SELECT * FROM table_a WHERE search="a-b-c"`, since we might alter the user query by doing this.

Looks like the [transpile user's statement](https://github.com/ploomber/jupysql/pull/164#discussion_r1122076690) might be the ultimate way, somehow it still makes sense since if the user is using PostgreSQL to write PostgreSQL statement, all we do just transpile the statement again to PostgreSQL fashion to ensure it's compatible

**Comment by tonykploomber** [comment]
> ```
> Tried to workaround
> 
> I tried to use [to_identifier](https://sqlglot.com/sqlglot/expressions.html#to_identifier) to quote the alias, but it will fail on mysql and mariadb, since those database doesn't support the double quote query like 'WITH "with-q" AS (SELECT * FROM test) SELECT * FROM sub_query_a'
> ```
> 
> you're probably not doing something right. if you specify mysql / mariadb, then it will translate the double quote to a back tick.

Assign the dialect parameter is doable, but when it comes to 

The challenge we faced here is we might expect the user provides `SELECT * FROM with-q`, but `sqlglot.parse_one("SELECT * FROM with-q")` is throwing some error, 

Quote the table name is fine according to your suggestion! But somehow we need the more robust way to auto-quote the thing, since we don't want to replace every `-` with `_` in the user's query

`sqlglot.parse_one('SELECT * FROM "with-q"')` - this works fine

**Comment by tobymao** [comment]
> 

i would just not allow "select * from with-q" as valid input. it's non standard sql.

**Comment by edublancas** [comment]
ok, sounds like the dashes are creating a lot of trouble, so let's move forward with your original suggestion @tonykploomber and do not allow them; since we'll break compatibility, let's display a message saying how to fix it:

> dashes are now allowed, replace them with an underscore: no_nulls


@tobymao thanks so much for your help!

**Comment by tobymao** [comment]
> ok, sounds like the dashes are creating a lot of trouble, so let's move forward with your original suggestion @tonykploomber and do not allow them; since we'll break compatibility, let's display a message saying how to fix it:
> 
> > dashes are now allowed, replace them with an underscore: no_nulls
> 
> @tobymao thanks so much for your help!

you can also tell your users to explicitly quote these things in their sql

**Comment by tonykploomber** [comment]
I decide to make a new PR...just the testing environment is not stable now, probably I introduced this

**Comment by tonykploomber** [comment]
After sqlglot updated to 3.7, the build issue was resolved

**Review by edublancas — CHANGES_REQUESTED** [review]
alright, we're getting closer!

a few things. I opened https://github.com/ploomber/jupysql/issues/245 so we make a release and give some time to users to update their code.

since this is a breaking API change, we need to update the `__version__` in the `__init__.py` file (0.7.0dev) and also create a new section in the changelog with such version and put the new changelog entry there

finally, we need to pin the version in setup.py, I'm guessing we need `sqlglot>=11.3.7`

https://github.com/tobymao/sqlglot/releases/tag/v11.3.7

**Comment by edublancas** [comment]
I'll make a release now that will include https://github.com/ploomber/jupysql/pull/249 - we can now delete the warnings we added there in this PR so please rebase.

also address the comments from my previous comment

**Comment by tonykploomber** [comment]
> I'll make a release now that will include #249 - we can now delete the warnings we added there in this PR and turn them into errors so please rebase.
> 
> also address the comments from my previous comment

1. Versions handled - [4b2caca](https://github.com/ploomber/jupysql/pull/164/commits/4b2caca4458c7f907ae42f77b770b79d8b11507a)
2. Delete the warning and now it will throw the error - [d17e5c3](https://github.com/ploomber/jupysql/pull/164/commits/d17e5c33202f094ae22dcba48d7c92e89654607d)

File: src/sql/store.py [hunk 11]
```diff
@@ -58,11 +58,10 @@ def render(self, query, with_=None):
     @modify_exceptions
     def store(self, key, query, with_=None):
         if "-" in key:
-            print(
-                "Dash is not suggested in standard sql clause,"
-                " please replace this with an underscore, e.g. no-nulls -> no_nulls"
+            raise SyntaxError(
```
Comment (edublancas) [inline src/sql/store.py:61]: this should be a UsageError (`from IPython.core.error import UsageError`)

**Review by edublancas — CHANGES_REQUESTED** [review]
added a comment and fixed some merge conflicts

File: src/sql/store.py (hunk 11 above)
Comment (tonykploomber) [reply src/sql/store.py]: Fix in [d50f586](https://github.com/ploomber/jupysql/pull/164/commits/d50f58667efbc07e8579ee735b6c0e4c7094c596)

Resolved conflict

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
awesome! 🎉🎉
